pkg/vsphere/guestinfo/encode.go

```diff
@@ -0,0 +1,291 @@
+	//log.SetLevel(log.DebugLevel)
+
+	return EncodeWithPrefix(i, DefaultPrefix)
+}
+
+// EncodeWithPrefix convert given type to []types.BaseOptionValue
+func EncodeWithPrefix(i interface{}, prefix string) []types.BaseOptionValue {
+	var config []types.BaseOptionValue
+
+	// value representing the run-time data
+	value := reflect.ValueOf(i)
+	log.Debugf("Value: %#v", value)
+
+	// determine the kind as it changes how we get underlying data
+	switch value.Kind() {
+	case reflect.Invalid:
+		log.Errorf("Invalid Kind: %#v", value)
+		return nil
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Bool, reflect.String, reflect.Float32, reflect.Float64:
+		// handle basic types directly
+		log.Debugf("Basic type: %s", prefix)
+
+		return append(config, &types.OptionValue{Key: prefix, Value: ToString(value)})
+	case reflect.Slice, reflect.Array:
+		log.Debugf("Slice BEGIN: %s", prefix)
+
+		var values []string
+		// iterate over slice to find what's underneath
+		for i := 0; i < value.Len(); i++ {
+			// recurse if we have a struct type
+			// otherwise add it's value to values
+			if value.Index(i).Kind() == reflect.Struct {
+				key := fmt.Sprintf("%s|%d", prefix, i)
+				config = append(config, EncodeWithPrefix(value.Index(i).Interface(), key)...)
+			} else {
+				values = append(values, ToString(value.Index(i)))
+			}
+		}
+		// set the slice key with the values seperated by |
+		if len(values) > 0 {
+			// sort the values before joining
+			sort.Strings(values)
+			// prefix~ contains the items
+			config = append(config, &types.OptionValue{Key: fmt.Sprintf("%s~", prefix), Value: strings.Join(values, "|")})
+			// prefix contains the count
+			config = append(config, &types.OptionValue{Key: prefix, Value: fmt.Sprintf("%d", value.Len()-1)})
+		} else {
+			v := reflect.ValueOf(value.Len() - 1)
+			config = append(config, &types.OptionValue{Key: prefix, Value: ToString(v)})
+		}
+		log.Debugf("Slice END %s %s", prefix, values)
+
+		return config
+
+	case reflect.Struct:
+		// iterate through every struct type field
+		for i := 0; i < value.NumField(); i++ {
+
+			structField := value.Type().Field(i)
+			//skip unexported fields
+			if structField.PkgPath != "" {
+				log.Debugf("Skipping %s (not exported)", structField.Name)
+				continue
+			}
+
+			// get the annotations
+			tags := structField.Tag
+			log.Debugf("Tags: %#v", tags)
+
+			// do we have DefaultTagName?
+			if tags.Get(DefaultTagName) == "" {
+				log.Debugf("Skipping %s (not vic)", structField.Name)
+				continue
+			}
+
+			// get the scopes
+			scopes := strings.Split(tags.Get("scope"), ",")
+
+			// get the keys and split properties from it
+			keys := strings.Split(tags.Get("key"), ",")
+			key, properties := keys[0], keys[1:]
+
+			// returns the i'th field of the struct i
+			field := value.Field(i)
+
+			// process properties
+			if len(properties) > 0 {
+				// do not recurse into nested type
+				if properties[0] == "omitnested" {
```

Comment:
Yep will try to write a short something to document the syntax, our assumptions and rules